File: Assignment3/code1.py
#%%
import numpy as np

def gemm_ijk(alpha, A, B, beta, C):
    assert A.shape[1] == B.shape[0]
    assert A.shape[0] == C.shape[0]
    assert B.shape[1] == C.shape[1]
    m, p = A.shape
    n = B.shape[1]
    for i in range(m):
        for j in range(n):
            sum_ = 0.0
            for k in range(p):
                sum_ += A[i, k] * B[k, j]
            C[i, j] = alpha * sum_ + beta * C[i, j]

def gemm_ikj(alpha, A, B, beta, C):
    assert A.shape[1] == B.shape[0]
    assert A.shape[0] == C.shape[0]
    assert B.shape[1] == C.shape[1]
    m, p = A.shape
    n = B.shape[1]
    for i in range(m):
        for k in range(p):
            temp = alpha * A[i, k]
            for j in range(n):
                if k==0 :
                    C[i, j] *= beta
                C[i, j] += temp * B[k, j]



def gemm_jki(alpha, A, B, beta, C):
    assert A.shape[1] == B.shape[0]
    assert A.shape[0] == C.shape[0]
    assert B.shape[1] == C.shape[1]
    m, p = A.shape
    n = B.shape[1]
    for j in range(n):
        for k in range(p):
            temp = alpha * B[k, j]
            for i in range(m):
                if k==0 :
                    C[i, j] *= beta
                C[i, j] += temp * A[i, k]

def gemm_kij(alpha, A, B, beta, C):
    assert A.shape[1] == B.shape[0]
    assert A.shape[0] == C.shape[0]
    assert B.shape[1] == C.shape[1]
    m, p = A.shape
    n = B.shape[1]
    for k in range(p):
        for i in range(m):
            temp = alpha * A[i, k]
            for j in range(n):
                if k==0 :
                    C[i, j] *= beta
                C[i, j] += temp * B[k, j]


def gemm_kji(alpha, A, B, beta, C):
    assert A.shape[1] == B.shape[0]
    assert A.shape[0] == C.shape[0]
    assert B.shape[1] == C.shape[1]
    m, p = A.shape
    n = B.shape[1]
    for k in range(p):
        for j in range(n):
            temp = alpha * B[k, j]
            for i in range(m):
                if k==0 :
                    C[i, j] *= beta
                C[i, j] += temp * A[i, k]

# Example usage
A = np.random.rand(2, 3)
B = np.random.rand(3, 4)
c = np.random.rand(2, 4)
alpha = 1.0
beta = 0.0
C_dgemm = alpha * np.dot(A, B) + beta * c

C = c.copy()
gemm_ijk(alpha, A, B, beta, C)
print("C (ijk):")
print(C)
print("Are the results equal?", np.allclose(C, C_dgemm))


C = c.copy()
gemm_ikj(alpha, A, B, beta, C)
print("C (ikj):")
print(C)
print("Are the results equal?", np.allclose(C, C_dgemm))


C = c.copy()
gemm_jki(alpha, A, B, beta, C)
print("C (jki):")
print(C)
print("Are the results equal?", np.allclose(C, C_dgemm))


C = c.copy()
gemm_kij(alpha, A, B, beta, C)
print("C (kij):")
print(C)
print("Are the results equal?", np.allclose(C, C_dgemm))


C = c.copy()
gemm_kji(alpha, A, B, beta, C)
print("C (kji):")
print(C)
print("Are the results equal?", np.allclose(C, C_dgemm))






m = 3
n = 4
p = 2
alpha = 0.0
beta = 2.0

A = np.array([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]]).T
B = np.array([[7.0, 8.0], [9.0, 10.0], [11.0, 12.0], [13.0, 14.0]]).T
C = np.array([[1,2,3,4],[4,5,6,7],[8,9,10,11]])
C_dgemm = alpha * np.dot(A, B) + beta * C
print("C_dgemm:")
print(C_dgemm)




# %%

File: Assignment3/code1.cpp
#include <iostream>
#include <vector>

template <typename T>
void gemm_ijk(const int m, const int n, const int p, const T alpha, const T beta, const std::vector<T>& A, const std::vector<T>& B, std::vector<T>& C) {
    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < n; ++j) {
            T sum = 0.0;
            for (int k = 0; k < p; ++k) {
                sum += A[k * m + i] * B[j * p + k];
            }
            C[j * m + i] = alpha * sum + beta * C[j * m + i];
        }
    }
}

int main() {
    const int m = 3;
    const int n = 4;
    const int p = 2;
    const double alpha = 3.1;
    const double beta = 2.4;

    std::vector<double> A = {1.0, 2.0, 3.0, 4.0, 5.0, 6.0};
    std::vector<double> B = {7.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0};
    std::vector<double> C = {1.0, 4.0, 8.0, 2.0, 5.0, 9.0, 3.0, 6.0, 10.0, 4.0, 7.0, 11.0 };

    gemm_ijk(m, n, p, alpha, beta, A, B, C);

    std::cout << "C = ";
    for (int i = 0; i < m * n; ++i) {
        std::cout << C[i] << " ";
    }
    std::cout << std::endl;

    return 0;
}